qemu/tests/block_resize.py

```diff
@@ -9,6 +9,166 @@
+    """
+    Get the drive id in linux guest.
+
+    param data_image: data image name, like image1 or stg.
+    param return: return drive path, like /dev/sda or /dev/sdb.
+    """
+    pattern = r"(serial|wwn)=(\w+)"
+    match = re.search(pattern, params["blk_extra_params_%s" % data_image], re.M)
+    if match:
+        drive_id = match.group(2)
+    else:
+        test.fail("No available tag to get drive id")
+    drive_path = utils_misc.get_linux_drive_path(session, drive_id)
+    if not drive_path:
+        test.error("Failed to get '%s' drive path" % data_image)
+    return drive_path
+
+
+def block_resize_enlarge(vm, session, params, device, block_size):
+    """
+    Enlarge the disk image of device to expected block_size.
+
+    param device: enlarge device.
+    param block_size: enlarge device to block_size.
+    """
+    logging.info("Enlarge disk size to %s in monitor"
+                 % block_size, logging.info)
+    vm.monitor.block_resize(device, block_size)
+    pattern = params.get("block_size_pattern")
+    block_size_cmd = params["block_size_cmd"]
+    if params.get("os_type") == "windows":
+        utils_misc.extend_shrink_windows_disk(session, params["disk_index"],
+                                              params["disk_letter"],
+                                              operation="extend")
+
+
+def block_resize_shrink(vm, session, params, device, old_disk_size, shrunk_size):
+    """
+    Shrink the disk image of device to expected block_size.
+
+    param device: shrink device.
+    param old_disk_size: the size before shrink.
+    param shrunk_size: want to shrink size, after shrunk, the disk size
+    should equal to "old_disk_size - shrunk_size".
+
+    param return: return the block_size: "old_disk_size-shrunk_size".
+    """
+    if params.get("os_type") == "windows":
+        shrunk_size = utils_misc.extend_shrink_windows_disk(session,
+                                                            params["disk_index"],
+                                                            params["disk_letter"],
+                                                            operation="shrink")
+        shrunk_size = float(utils_misc.normalize_data_size("%sM" % shrunk_size,
+                                                           order_magnitude="B"))
+    block_size = old_disk_size - int(shrunk_size)
+    logging.info("Shrink disk size to %s in monitor"
+                 % block_size, logging.info)
+    vm.monitor.block_resize(device, block_size)
+    return block_size
+
+
+def rescan_disk_after_resize(vm, session, params):
+    """
+    Rescan disk in windows or linux guest.
+    In linux guest, after block_resize:
+    need to rescan for virtio-scsi disk(no need reboot),
+    need to reboot for virtio-blk disk(no need rescan).
+    In windows guest, agter block_resize:
+    need to rescan for both virtio-scsi and virtio-blk disk.
+    """
+    if params.get("guest_prepare_cmd"):
```

Comment:
Has no relationship with rescan, so doubt it's convenient to do operations of "guest_prepare_cmd" and "need_reboot" cmd here.

Reply:
as our discuss, will change the function name to refresh_disk_after_resize() here.
thanks a lot~